// ngx-websocket/esm5/lib/ngx-websocket.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component } from '@angular/core';
import { NgxWebsocketService } from './ngx-websocket.service';
var NgxWebsocketComponent = /** @class */ (function () {
    function NgxWebsocketComponent(websocketService) {
        this.websocketService = websocketService;
    }
    NgxWebsocketComponent.decorators = [
        { type: Component, args: [{
                    selector: 'enl-ngx-websocket',
                    template: "\n    <p>\n      ngx-websocket works!\n    </p>\n    <div>\n      <button>\u53D1\u9001</button>\n    </div>\n  ",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    NgxWebsocketComponent.ctorParameters = function () { return [
        { type: NgxWebsocketService }
    ]; };
    return NgxWebsocketComponent;
}());
export { NgxWebsocketComponent };
function NgxWebsocketComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    NgxWebsocketComponent.prototype.socket;
    /** @type {?} */
    NgxWebsocketComponent.prototype.websocketService;
}
