// ngx-websocket/esm5/lib/ngx-websocket.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Socket } from './socket';
import * as i0 from "@angular/core";
/**
 * this service provides multiple methods to create websocket, add callback function to websocekt or destory webscoket.
 */
var NgxWebsocketService = /** @class */ (function () {
    function NgxWebsocketService() {
        this.sockets = new Map();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    NgxWebsocketService.prototype.open = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        if (!this.checkExistence(url)) {
            var /** @type {?} */ socket = new Socket(url);
            this.sockets.set(url, socket);
            return socket;
        }
    };
    /**
     * @param {?} url
     * @return {?}
     */
    NgxWebsocketService.prototype.checkExistence = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.sockets.forEach(function (key, value) {
            if (key === url) {
                //check connection state
                console.warn("a websocket with same url has existed");
                return true;
            }
        });
        return false;
    };
    NgxWebsocketService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    NgxWebsocketService.ctorParameters = function () { return []; };
    /** @nocollapse */ NgxWebsocketService.ngInjectableDef = i0.defineInjectable({ factory: function NgxWebsocketService_Factory() { return new NgxWebsocketService(); }, token: NgxWebsocketService, providedIn: "root" });
    return NgxWebsocketService;
}());
export { NgxWebsocketService };
function NgxWebsocketService_tsickle_Closure_declarations() {
    /** @type {?} */
    NgxWebsocketService.prototype.sockets;
}
