// ngx-websocket/fesm2015/ngx-websocket.js
import { Injectable, Component, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class Subscribe {
    /**
     * @param {?} _type
     * @param {?} _f
     */
    constructor(_type, _f) {
        this.type = _type;
        this.f = _f;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Socket class
 * each socket handles a single WebScoket Connection.
 */
class Socket {
    /**
     * @param {?} url
     */
    constructor(url) {
        this.url = url;
        try {
            this.websocket = new WebSocket(url);
        }
        catch (/** @type {?} */ e) {
            console.error(e);
        }
        this._state = "closed";
        this._subscribes = [];
        //onOpen
        this.websocket.addEventListener("open", (e) => {
            //websocket stream start
            this.subscribeLoop(true);
        });
        //onError
        this.websocket.addEventListener("error", (e) => {
            console.error("websocket onError");
        });
        //onMessage
        // this._messageObservable = Observable.create((observer) => {
        //     this.websocket.addEventListener("message", (e) => {
        //         observer.next({
        //             event: e
        //         });
        //         //stream start again
        //         this.subscribeLoop(false);
        //     });
        // });
        //onMessage
        this.websocket.addEventListener("message", (e) => {
            this._receiveData = e;
            //stream start again
            this.subscribeLoop(false);
        });
        //onClose
        this.websocket.addEventListener("close", (e) => {
            this._state = "closed";
            console.warn("websocket onClose");
        });
    }
    /**
     * reveive data from previous handler, wouldn't block current stream.
     * @param {?} f
     * @return {?}
     */
    then(f) {
        this._subscribes.push(new Subscribe(0, f));
        return this;
    }
    /**
     * callback funtion when websocket reveive message.
     * @param {?} f
     * @return {?}
     */
    message(f) {
        this._subscribes.push(new Subscribe(1, f));
        return this;
    }
    /**
     * catch error or close state on websocket;
     * @return {?}
     */
    catch() {
        return this;
    }
    /**
     * finally handler.
     * @return {?}
     */
    finally() {
        return this;
    }
    /**
     * send message via this websocket
     * @param {?} message
     * @return {?}
     */
    send(message) {
        try {
            this.websocket.send(message);
        }
        catch (/** @type {?} */ e) {
            console.error(e);
        }
        return this;
    }
    /**
     * close websocket
     * @return {?}
     */
    close() {
        this.websocket.close();
    }
    /**
     * traversal subscirbes set, execute action according to item's type
     * @param {?} init if subscribeLoop is the first tiem to execute, than every 'then' handler before the first message handler will be remove after execute.
     * @return {?}
     */
    subscribeLoop(init) {
        for (let /** @type {?} */ i = 0, /** @type {?} */ len = this._subscribes.length; i < len; i++) {
            let /** @type {?} */ sub = this._subscribes[i];
            if (sub.type === 0) {
                this._passingData = sub.f(this._passingData, this);
                if (init === true)
                    this._subscribes[i] = new Subscribe(0, () => { });
            }
            if (sub.type === 1) {
                if (init === true)
                    break;
                // this._messageObservable.subscribe({
                //     next(d) {
                //         socket._passingData = sub.f((d.event.data), socket._passingData, socket, d.event);
                //     },
                //     error(msg) {
                //         //call the catch handler
                //     }
                // });
                this._passingData = sub.f(this._receiveData.data, this._passingData, this, this._receiveData);
            }
        }
    }
    /**
     * @return {?}
     */
    get state() {
        return this._state;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * this service provides multiple methods to create websocket, add callback function to websocekt or destory webscoket.
 */
class NgxWebsocketService {
    constructor() {
        this.sockets = new Map();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    open(url) {
        if (!this.checkExistence(url)) {
            let /** @type {?} */ socket = new Socket(url);
            this.sockets.set(url, socket);
            return socket;
        }
    }
    /**
     * @param {?} url
     * @return {?}
     */
    checkExistence(url) {
        this.sockets.forEach((key, value) => {
            if (key === url) {
                //check connection state
                console.warn("a websocket with same url has existed");
                return true;
            }
        });
        return false;
    }
}
NgxWebsocketService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
NgxWebsocketService.ctorParameters = () => [];
/** @nocollapse */ NgxWebsocketService.ngInjectableDef = defineInjectable({ factory: function NgxWebsocketService_Factory() { return new NgxWebsocketService(); }, token: NgxWebsocketService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxWebsocketComponent {
    /**
     * @param {?} websocketService
     */
    constructor(websocketService) {
        this.websocketService = websocketService;
    }
}
NgxWebsocketComponent.decorators = [
    { type: Component, args: [{
                selector: 'enl-ngx-websocket',
                template: `
    <p>
      ngx-websocket works!
    </p>
    <div>
      <button>发送</button>
    </div>
  `,
                styles: []
            },] },
];
/** @nocollapse */
NgxWebsocketComponent.ctorParameters = () => [
    { type: NgxWebsocketService }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxWebsocketModule {
}
NgxWebsocketModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [NgxWebsocketComponent],
                exports: [NgxWebsocketComponent]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { NgxWebsocketService, NgxWebsocketComponent, NgxWebsocketModule, Socket };
