// ngx-websocket/esm2015/lib/ngx-websocket.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Socket } from './socket';
import * as i0 from "@angular/core";
/**
 * this service provides multiple methods to create websocket, add callback function to websocekt or destory webscoket.
 */
export class NgxWebsocketService {
    constructor() {
        this.sockets = new Map();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    open(url) {
        if (!this.checkExistence(url)) {
            let /** @type {?} */ socket = new Socket(url);
            this.sockets.set(url, socket);
            return socket;
        }
    }
    /**
     * @param {?} url
     * @return {?}
     */
    checkExistence(url) {
        this.sockets.forEach((key, value) => {
            if (key === url) {
                //check connection state
                console.warn("a websocket with same url has existed");
                return true;
            }
        });
        return false;
    }
}
NgxWebsocketService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
NgxWebsocketService.ctorParameters = () => [];
/** @nocollapse */ NgxWebsocketService.ngInjectableDef = i0.defineInjectable({ factory: function NgxWebsocketService_Factory() { return new NgxWebsocketService(); }, token: NgxWebsocketService, providedIn: "root" });
function NgxWebsocketService_tsickle_Closure_declarations() {
    /** @type {?} */
    NgxWebsocketService.prototype.sockets;
}
